### PR TITLE
Fix Unicode Input Issue and OSError in Test Script

### PR DESCRIPTION
**Pull request by sakkyoi**

### CODE REVIEW

File: readchar/_win_read.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def readkey() -> str:
         raise KeyboardInterrupt
 
     # if it is a normal character:
-    if ch not in "\x00\xe0":
+    if ch not in "\x00\xc3\xa0":
```
Comment (Cube707) [inline readchar/_win_read.py:24]: what is the reason for this change?

only `0x00` and `0xe0` are documented prefix chars that mark a multi-char sequence, so what are the other chars?

File: readchar/_win_read.py [hunk 2]
```diff
@@ -7,8 +7,8 @@ def readchar() -> str:
     """Reads a single character from the input stream.
```
Comment (Cube707) [inline readchar/_win_read.py:7]:
```suggestion
    """Reads a single utf8-character from the input stream.
```

File: tests/windows/conftest.py [hunk 3]
```diff
@@ -14,10 +14,16 @@ def pytest_ignore_collect(path, config):
...
         def push(self, string):
-            for c in string:
-                msvcrt.ungetch(ord(c).to_bytes(1, "big"))
+            # Create an iterator from the string
+            characters = iter(string)
+
+            # Patch msvcrt.getwch to return the next character from the iterator.
+            # Don't use next(iter(string)) as it creates a new iterator each time.
+            # For example,
+            # if you use a new iterator, cursor_up (\x00\x48) will be read as \x00\x00.
```
Comment (Cube707) [inline tests/windows/conftest.py:26]:
```suggestion
            # Patch msvcrt.getwch to return the next character from the iterator.
```

File: tests/windows/conftest.py [hunk 4]
```diff
@@ -3,8 +3,8 @@
 import pytest
 
 
-if sys.platform in ("win32", "cygwin"):
-    import msvcrt
+# if sys.platform in ("win32", "cygwin"):
+#     import msvcrt
```
Comment (Cube707) [inline tests/windows/conftest.py:7]: don't just comment out stuff and leaf dead code in the file. If its no longer needed, remove it.
```suggestion
```

**Review by Cube707 — COMMENTED** [review]
those are my imidieate thought on the changes.

Also, how this this new behavior compare to Linux? can all wide-charcters allready be read on linux systems or must changes be also implemented there?

File: readchar/_win_read.py (hunk 1 above)
Comment (sakkyoi) [reply readchar/_win_read.py:24]: The reason is that when reading with msvcrt.getwch(), \xe0 is read as \xc3\xa0.

**Comment by sakkyoi** [comment]
I tested it on macOS without any issues. The original problem seems to occur only on Windows.

**Comment by Cube707** [comment]
So you tested with relevant characters on MacOS?
But not on Windows?

**Comment by sakkyoi** [comment]
I apologize for my incomplete explanation. I initially noticed this issue on Windows, so I fixed it. However, I also noticed that Windows handles this differently compared to Linux and macOS. I tested it on macOS, and it showed that macOS can natively handle Unicode character input perfectly.

**Comment by Cube707** [comment]
Took me a while to get the Japanese keyboard to work on windows, but here is what I found:

This seems to mostly work as expected. Linux is fine as is, as you have found with Mac:
grafik

windows work for all characters I tested, but breaks on emojis for some reason:
grafik

File: readchar/_win_read.py (hunk 1 above)
Comment (sakkyoi) [reply readchar/_win_read.py]: Oh, I see where the problem is. At first, I tried converting it to UTF-8 for detection(like `if ch.encode("utf-8") not in not in b"\x00\xc3\xa0":`), but then I reverted it to the original implementation (like `if ch not in "\x00\xe0":`). This should be changed back.

**Comment by sakkyoi** [comment]
After my investigation, it appears that some emojis are encoded using UTF-8 surrogates, which may appear as two characters in the terminal (you can observe this if you select them with your cursor). This is related to the issue of "surrogates not allowed." Perhaps you could use a try-except block to handle UnicodeEncodeError, as this error is raised by getwchar().

**Comment by sakkyoi** [comment]
I might have found a solution, so please stay tuned.

**Comment by Cube707** [comment]
> After my investigation, it appears that some emojis are encoded using UTF-8 surrogates, which may appear as two characters in the terminal 

Still strang that it works fine on linux

**Comment by sakkyoi** [comment]
Linux works fine without this issue.

**Comment by Cube707** [comment]
this seems to be a good solution, but I would prefer that code in the `readkey()` function, as it handles identically to the combined key-presses. Surrogates are also made up of two chars that don't make sense on their own.

I will implement that now

**Review by Cube707 — APPROVED** [review]
